vm/src/rv32/parse.rs

```diff
@@ -119,17 +119,35 @@ fn sop(word: u32) -> Option<SOP> {
 // parsing of arithmetic operations pg. 18-20
 
 fn aop(word: u32) -> Option<AOP> {
-    let res = match (opcode(word), funct3(word), funct7(word)) {
-        (0b0110011, 0b000, 0b0100000) => SUB,
-        (_, 0b000, _) => ADD,
-        (_, 0b001, _) => SLL,
-        (_, 0b010, _) => SLT,
-        (_, 0b011, _) => SLTU,
-        (_, 0b100, _) => XOR,
-        (_, 0b101, 0b0000000) => SRL,
-        (_, 0b101, 0b0100000) => SRA,
-        (_, 0b110, _) => OR,
-        (_, 0b111, _) => AND,
+    assert_eq!(opcode(word), OPC_ALU);
+    let res = match (funct3(word), funct7(word)) {
+        (0b000, 0b0100000) => SUB,
+        (0b000, 0b0000000) => ADD,
+        (0b001, 0b0000000) => SLL,
+        (0b010, 0b0000000) => SLT,
+        (0b011, 0b0000000) => SLTU,
+        (0b100, 0b0000000) => XOR,
+        (0b101, 0b0000000) => SRL,
+        (0b101, 0b0100000) => SRA,
+        (0b110, 0b0000000) => OR,
+        (0b111, 0b0000000) => AND,
+        _ => return None,
+    };
+    Some(res)
+}
+
+fn aopi(word: u32) -> Option<AOP> {
+    assert_eq!(opcode(word), OPC_ALUI);
+    let res = match (funct3(word), funct7(word) /* imm[11:5] */) {
+        (0b000, _ /* imm[11:5] can be any */) => ADD,
+        (0b001, 0b0000000 /* shift amount is required to be small */) => SLL,
+        (0b010, _ /* imm[11:5] can be any */) => SLT,
+        (0b011, _ /* imm[11:5] can be any */) => SLTU,
+        (0b100, _ /* imm[11:5] can be any */) => XOR,
+        (0b101, 0b0000000 /* shift amount is required to be small */) => SRL,
```

Comment:
Also I'm confused by the "small" part, this value doesn't have anything to do with the shift amount, right? It's a fixed value to indicate the type of shift, I think.

Reply:
I can argue both ways, but I'll just say something else, like, SLLI requires special value here.

Reply:
done.